File: packages/indexer/indexer/static/syntax-record/tree-paths.ts
import type { ProjectDefinition, ProjectDefinitionKind } from '@crux/core/project-index'
import type { IndexerExtensionRuntime } from '../../extensions'
import { staticFoundDefinitionsFromExtractedFacts } from '../../extensions/static-normalizer'
import type { StaticFoundDefinition } from '../../types'
import type { StaticSyntaxFileRecord, StaticSyntaxValue } from './types'
import { createNativeFactIndex, extractedFacts } from './native-facts'

/** Runtime inputs for record-backed prompt/context tree path projection. */
export interface StaticRecordTreePathInput {
  /** Project root used by extractor runtime calls. */
  readonly root: string
  /** Source record that owns the authored tree calls. */
  readonly record: StaticSyntaxFileRecord
  /** Runtime used to extract imported tree leaf definitions through the same extension contract. */
  readonly runtime: IndexerExtensionRuntime
  /** Definitions already discovered from this source file, keyed by authored variable names. */
  readonly found: readonly StaticFoundDefinition[]
  /** Reads a syntax record for an imported source file. */
  readonly readRecord: (file: string) => Promise<StaticSyntaxFileRecord>
}

/**
 * Projects `createPrompts` and `createContexts` object trees from syntax records.
 *
 * The projection intentionally consumes the compact record ABI instead of parser-native AST nodes.
 * Only identifier leaves are resolved, matching the conservative TypeScript projection semantics.
 */
export async function staticRecordTreePathDefinitions(input: StaticRecordTreePathInput): Promise<ProjectDefinition[]> {
  const definitions: ProjectDefinition[] = []
  const localByVariable = new Map(input.found.map((item) => [item.variableName, item.definition]))

  for (const tree of treeContainers(input.record)) {
    definitions.push(
      ...(await treePathDefinitionsForObject({
        input,
        object: tree.object,
        path: [],
        kind: tree.kind,
        localByVariable,
      })),
    )
  }

  return definitions
}

interface TreeContainer {
  readonly kind: Extract<ProjectDefinitionKind, 'prompt' | 'context'>
  readonly object: Extract<StaticSyntaxValue, { readonly kind: 'object' }>
}

function treeContainers(record: StaticSyntaxFileRecord): readonly TreeContainer[] {
  return record.localInitializers.flatMap((initializer) => treeContainerFromValue(initializer.value))
}

function treeContainerFromValue(value: StaticSyntaxValue): readonly TreeContainer[] {
  if (value.kind !== 'call') return []
  const kind = treeKindForCall(value.callee.name)
  const firstArg = value.args[0]
  if (!kind || firstArg?.kind !== 'object') return []
  return [{ kind, object: firstArg }]
}

async function treePathDefinitionsForObject(input: {
  readonly input: StaticRecordTreePathInput
  readonly object: Extract<StaticSyntaxValue, { readonly kind: 'object' }>
  readonly path: readonly string[]
  readonly kind: Extract<ProjectDefinitionKind, 'prompt' | 'context'>
  readonly localByVariable: ReadonlyMap<string, ProjectDefinition>
}): Promise<ProjectDefinition[]> {
  const definitions: ProjectDefinition[] = []
  for (const property of input.object.properties) {
    if (property.spread) continue
    const nextPath = [...input.path, property.name]
    if (property.value.kind === 'object') {
      definitions.push(
        ...(await treePathDefinitionsForObject({
          ...input,
          object: property.value,
          path: nextPath,
        })),
      )
      continue
    }
    if (property.value.kind !== 'identifier') continue

    const resolved = await resolveTreeLeafDefinition({
      input: input.input,
      localByVariable: input.localByVariable,
      identifier: property.value.name,
      kind: input.kind,
    })
    if (!resolved) continue
    definitions.push({
      id: resolved.id,
      kind: resolved.kind,
      name: resolved.name,
      path: nextPath,
      fidelity: resolved.fidelity,
      status: resolved.status,
    })
  }
  return definitions
}

async function resolveTreeLeafDefinition(input: {
  readonly input: StaticRecordTreePathInput
  readonly localByVariable: ReadonlyMap<string, ProjectDefinition>
  readonly identifier: string
  readonly kind: ProjectDefinitionKind
}): Promise<ProjectDefinition | undefined> {
  const local = input.localByVariable.get(input.identifier)
  if (local?.kind === input.kind) return local

  const importRecord = input.input.record.imports.find((item) => item.localName === input.identifier)
  if (!importRecord?.resolvedFile || importRecord.importedName === 'default') return undefined
  const exported = await staticExportDefinitions(input.input, importRecord.resolvedFile)
  const imported = exported.get(importRecord.importedName)
  return imported?.kind === input.kind ? imported : undefined
}

async function staticExportDefinitions(
  input: StaticRecordTreePathInput,
  file: string,
): Promise<Map<string, ProjectDefinition>> {
  const record = await safeReadRecord(input.readRecord, file)
  if (!record) return new Map()
  const definitions = new Map<string, ProjectDefinition>()
  const nativeFactsByMatchIndex = createNativeFactIndex(record)

  for (const [matchIndex, match] of record.matches.entries()) {
    if (!match.exported) continue
    const nativeFacts = nativeFactsByMatchIndex.get(matchIndex)
    const extracted = [
      ...(nativeFacts?.facts ?? []),
      ...extractedFacts(
        input.runtime.extractStaticRecord({
          root: input.root,
          record,
          match,
          skipExtractors: nativeFacts?.replacedExtractors,
        }),
      ),
    ]
    for (const found of staticFoundDefinitionsFromExtractedFacts(extracted)) {
      definitions.set(found.variableName, found.definition)
    }
  }

  return definitions
}

async function safeReadRecord(
  readRecord: (file: string) => Promise<StaticSyntaxFileRecord>,
  file: string,
): Promise<StaticSyntaxFileRecord | undefined> {
  try {
    return await readRecord(file)
  } catch {
    return undefined
  }
}

function treeKindForCall(name: string): Extract<ProjectDefinitionKind, 'prompt' | 'context'> | undefined {
  if (name === 'createPrompts') return 'prompt'
  if (name === 'createContexts') return 'context'
  return undefined
}

File: crates/static-compiler/src/source/tree_paths.rs
//! Record-backed prompt/context tree path projection for Static Index analyze.
//!
//! The TypeScript compiler adds path overlays from `createPrompts` and `createContexts` trees
//! after primitive extraction. Native analyze emits the same minimal duplicate definitions and lets
//! finalization merge them into the richer primary prompt/context rows.

use std::collections::HashMap;

use serde_json::{Map, Value, json};

use crate::protocol::{
    StaticInitializerRecord, StaticNativeFactProjection, StaticSourceMatch, StaticSyntaxFileRecord,
    StaticSyntaxValue,
};

pub(crate) fn grouped_tree_path_definition_facts(
    root: &str,
    project_name: Option<&str>,
    record: &StaticSyntaxFileRecord,
    native_facts: &[StaticNativeFactProjection],
    records_by_file: &HashMap<String, StaticSyntaxFileRecord>,
) -> Option<Value> {
    let local = definitions_by_variable(native_facts);
    let mut definitions = Vec::new();
    for initializer in &record.local_initializers {
        let Some((kind, object)) = tree_container(initializer) else {
            continue;
        };
        append_path_definitions(
            &mut definitions,
            TreeWalk {
                kind,
                object,
                path: Vec::new(),
                local: &local,
                record,
                records_by_file,
            },
        );
    }

    if definitions.is_empty() {
        return None;
    }

    let mut group = Map::new();
    group.insert("root".to_string(), Value::String(root.to_string()));
    if let Some(project_name) = project_name {
        group.insert(
            "projectName".to_string(),
            Value::String(project_name.to_string()),
        );
    }
    group.insert("definitions".to_string(), Value::Array(definitions));
    Some(Value::Object(group))
}

struct TreeWalk<'a> {
    kind: &'static str,
    object: &'a StaticSyntaxValue,
    path: Vec<String>,
    local: &'a HashMap<String, Value>,
    record: &'a StaticSyntaxFileRecord,
    records_by_file: &'a HashMap<String, StaticSyntaxFileRecord>,
}

fn append_path_definitions(definitions: &mut Vec<Value>, input: TreeWalk<'_>) {
    let StaticSyntaxValue::Object { properties, .. } = input.object else {
        return;
    };
    for property in properties
        .iter()
        .filter(|property| property.spread != Some(true))
    {
        let mut next_path = input.path.clone();
        next_path.push(property.name.clone());
        if matches!(property.value, StaticSyntaxValue::Object { .. }) {
            append_path_definitions(
                definitions,
                TreeWalk {
                    object: &property.value,
                    path: next_path,
                    ..input
                },
            );
            continue;
        }

        let StaticSyntaxValue::Identifier { name } = &property.value else {
            continue;
        };
        if let Some(definition) = resolve_tree_leaf_definition(
            name,
            input.kind,
            input.local,
            input.record,
            input.records_by_file,
        ) {
            definitions.push(path_overlay_definition(&definition, next_path));
        }
    }
}

fn tree_container(
    initializer: &StaticInitializerRecord,
) -> Option<(&'static str, &StaticSyntaxValue)> {
    let StaticSyntaxValue::Call { callee, args, .. } = &initializer.value else {
        return None;
    };
    let first_arg = args.first()?;
    let kind = match callee.name.as_str() {
        "createPrompts" => "prompt",
        "createContexts" => "context",
        _ => return None,
    };
    matches!(first_arg, StaticSyntaxValue::Object { .. }).then_some((kind, first_arg))
}

fn resolve_tree_leaf_definition(
    identifier: &str,
    kind: &str,
    local: &HashMap<String, Value>,
    record: &StaticSyntaxFileRecord,
    records_by_file: &HashMap<String, StaticSyntaxFileRecord>,
) -> Option<Value> {
    if let Some(definition) = local
        .get(identifier)
        .filter(|definition| definition_kind(definition) == Some(kind))
    {
        return Some(definition.clone());
    }

    let import = record
        .imports
        .iter()
        .find(|import| import.local_name == identifier)?;
    if import.imported_name == "default" {
        return None;
    }
    let imported_record = records_by_file.get(import.resolved_file.as_ref()?)?;
    exported_definitions_by_variable(imported_record)
        .get(&import.imported_name)
        .filter(|definition| definition_kind(definition) == Some(kind))
        .cloned()
}

fn definitions_by_variable(native_facts: &[StaticNativeFactProjection]) -> HashMap<String, Value> {
    native_facts
        .iter()
        .flat_map(|projection| fact_definitions_by_variable(&projection.facts))
        .collect()
}

fn exported_definitions_by_variable(record: &StaticSyntaxFileRecord) -> HashMap<String, Value> {
    record
        .native_facts
        .iter()
        .filter(|projection| source_match_exported(record.matches.get(projection.match_index)))
        .flat_map(|projection| fact_definitions_by_variable(&projection.facts))
        .collect()
}

fn fact_definitions_by_variable(facts: &Value) -> Vec<(String, Value)> {
    facts
        .get("definitions")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(|entry| {
            let variable_name = entry.get("variableName")?.as_str()?.to_string();
            let definition = entry.get("definition")?.clone();
            Some((variable_name, definition))
        })
        .collect()
}

fn path_overlay_definition(definition: &Value, path: Vec<String>) -> Value {
    json!({
        "id": definition["id"].clone(),
        "kind": definition["kind"].clone(),
        "name": definition["name"].clone(),
        "path": path,
        "fidelity": definition["fidelity"].clone(),
        "status": definition.get("status").cloned().unwrap_or(Value::Null),
    })
}

fn source_match_exported(source_match: Option<&StaticSourceMatch>) -> bool {
    match source_match {
        Some(StaticSourceMatch::Call { exported, .. })
        | Some(StaticSourceMatch::New { exported, .. })
        | Some(StaticSourceMatch::Object { exported, .. }) => *exported,
        None => false,
    }
}

fn definition_kind(definition: &Value) -> Option<&str> {
    definition.get("kind").and_then(Value::as_str)
}
